File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/utility/MarkdownGenerator.kt
package com.crescentflare.piratesgame.components.utility

import android.content.Context
import android.graphics.Typeface
import android.support.v4.content.ContextCompat
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.text.style.RelativeSizeSpan

import com.crescentflare.piratesgame.R
import com.crescentflare.piratesgame.components.styling.AppFonts
import com.crescentflare.simplemarkdownparser.core.MarkdownTag
import com.crescentflare.simplemarkdownparser.helper.AlignedListSpan
import com.crescentflare.simplemarkdownparser.helper.DefaultMarkdownSpanGenerator

/**
 * A utility to apply styles during markdown parsing
 */
class MarkdownGenerator @JvmOverloads constructor(private val applicationContext: Context, noColorization: Boolean = false, noBulletIndentation: Boolean = false) : DefaultMarkdownSpanGenerator() {

    // ---
    // Members
    // ---

    private var noColorization = false
    private var noBulletIndentation = false


    // ---
    // Initialization
    // ---

    init {
        this.noColorization = noColorization
        this.noBulletIndentation = noBulletIndentation
    }


    // ---
    // Implementation
    // ---

    override fun applySpan(builder: SpannableStringBuilder, type: MarkdownTag.Type, weight: Int, start: Int, end: Int, extra: String) {
        when (type) {
            MarkdownTag.Type.Header -> {
                if (!noColorization) {
                    builder.setSpan(
                        ForegroundColorSpan(
                            ContextCompat.getColor(
                                applicationContext,
                                R.color.colorPrimary
                            )
                        ), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
                builder.setSpan(
                    CustomTypefaceSpan(AppFonts.getTypeface("titleBold")),
                    start,
                    end,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                )
                if (weight == 1) {
                    builder.setSpan(
                        RelativeSizeSpan(
                            applicationContext.resources.getDimensionPixelSize(R.dimen.titleText).toFloat() / applicationContext.resources.getDimensionPixelSize(
                                R.dimen.text
                            ).toFloat()
                        ), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                } else if (weight == 2) {
                    builder.setSpan(
                        RelativeSizeSpan(
                            applicationContext.resources.getDimensionPixelSize(R.dimen.subTitleText).toFloat() / applicationContext.resources.getDimensionPixelSize(
                                R.dimen.text
                            ).toFloat()
                        ), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
                return
            }
            MarkdownTag.Type.TextStyle -> if (weight != 3) {
                val typeface = typefaceForWeight(weight)
                if (typeface != null) {
                    builder.setSpan(CustomTypefaceSpan(typeface), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                    return
                }
            }
            MarkdownTag.Type.OrderedList, MarkdownTag.Type.UnorderedList -> if (noBulletIndentation) {
                builder.setSpan(
                    AlignedListSpan(extra, 15 + (weight - 1) * 15, 5),
                    start,
                    end,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )
                return
            }
            else -> {}
        }
        super.applySpan(builder, type, weight, start, end, extra)
    }


    // ---
    // Helper
    // ---

    private fun typefaceForWeight(weight: Int): Typeface? {
        when (weight) {
            0 -> return AppFonts.getTypeface("normal")
            1 -> return AppFonts.getTypeface("italics")
            2 -> return AppFonts.getTypeface("bold")
        }
        return null
    }
}

File: pirates-ios/PiratesGame/Components/Utility/MarkdownGenerator.swift
//
//  MarkdownGenerator.swift
//  Component utility: apply styles during markdown parsing
//

import SimpleMarkdownParser

class MarkdownGenerator: DefaultMarkdownAttributedStringGenerator {
    
    // --
    // MARK: Members
    // --
    
    private var noColorization: Bool
    private var noBulletIndentation: Bool
    

    // --
    // MARK: Initialization
    // --
    
    init(noColorization: Bool = false, noBulletIndentation: Bool = false) {
        self.noColorization = noColorization
        self.noBulletIndentation = noBulletIndentation
    }
    

    // --
    // MARK: Implementation
    // --

    override func applyAttribute(defaultFont: UIFont, attributedString: NSMutableAttributedString, type: MarkdownTagType, weight: Int, start: Int, length: Int, extra: String) {
        switch type {
        case .header:
            var size = defaultFont.pointSize
            switch weight {
            case 1:
                size = AppDimensions.titleText
            case 2:
                size = AppDimensions.subTitleText
            default:
                break
            }
            if !noColorization {
                attributedString.addAttribute(NSAttributedString.Key.foregroundColor, value: AppColors.primary, range: NSMakeRange(start, length))
            }
            attributedString.addAttribute(NSAttributedString.Key.font, value: AppFonts.titleBold.font(ofSize: size), range: NSMakeRange(start, length))
            return
        case .textStyle:
            if let font = fontForWeight(weight: weight, pointSize: defaultFont.pointSize) {
                attributedString.addAttribute(NSAttributedString.Key.font, value: font, range: NSMakeRange(start, length))
                return
            }
            return
        case .orderedList, .unorderedList:
            if noBulletIndentation {
                let bulletParagraph = NSMutableParagraphStyle()
                let tokenTabStop = NSTextTab(textAlignment: .right, location: 10 + CGFloat(weight - 1) * 15, options: [:])
                let textTabStop = NSTextTab(textAlignment: .left, location: tokenTabStop.location + 5, options: [:])
                bulletParagraph.tabStops = [ tokenTabStop, textTabStop ]
                bulletParagraph.firstLineHeadIndent = 0
                bulletParagraph.headIndent = textTabStop.location
                attributedString.addAttribute(NSAttributedString.Key.paragraphStyle, value: bulletParagraph, range: NSMakeRange(start, length))
                return
            }
            break
        default:
            break
        }
        super.applyAttribute(defaultFont: defaultFont, attributedString: attributedString, type: type, weight: weight, start: start, length: length, extra: extra)
    }
    

    // --
    // MARK: Helper
    // --
    
    private func fontForWeight(weight: Int, pointSize: CGFloat) -> UIFont? {
        switch weight {
        case 0:
            return AppFonts.normal.font(ofSize: pointSize)
        case 1:
            return AppFonts.italics.font(ofSize: pointSize)
        case 2:
            return AppFonts.bold.font(ofSize: pointSize)
        case 3:
            return AppFonts.boldItalics.font(ofSize: pointSize)
        default:
            break
        }
        return nil
    }

}
